### PR TITLE
Feature: support for --empty=[drop|keep] in dolt_rebase()

### PR DESCRIPTION
**Pull request by fulghum**
Adds support for the `--empty` option to `dolt_rebase()`. This option controls how commits that **become** empty are handled. For example, if a branch is rebased and all the changes in one commit on that branch have already been applied to the upstream branch, then when that commit is reapplied, it will end up being empty. The two initially supported values are `drop` and `keep`. (Git also supports a `stop` value, which lets the user manually intervene.)

Also adds support for the `--allow-empty` flag for `dolt_cherry_pick()`. This flag controls whether Dolt will cherry-pick empty commits (i.e. commits that start off as empty, not commits that become empty after they are applied). 

These behaviors are slightly confusing for two reasons: 1) Git distinguishes between a commit that starts off empty and a commit that becomes empty while applying its changes, and 2) rebase and cherry-pick have slightly different default values for these two options. The differences are summarized below.

|  | Commits that start empty | Commits that become empty |
|-----| --------------------------| ------------------------------|
| `rebase` | default: `keep`, can be overridden with `--no-keep-empty` (`--keep-empty` is also supported) | default: `drop`, can be overridden with `--empty=keep`. For interactive rebases, the default changes to `stop`, which is not supported by Dolt yet.  |
| `cherry-pick` | default: `fail`, can be overridden with `--allow-empty` | default: `stop`, can be overridden with `--empty=keep` or `--empty=drop`. Dolt does not support `stop` yet, so Dolt's default is to fail.  |

Related issue: https://github.com/dolthub/dolt/issues/7830

### CODE REVIEW

**Comment by coffeegoddd** [comment]
@fulghum __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>4f3298a</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>4f3298a</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@fulghum __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>cda8661</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>cda8661</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@fulghum __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>b6785dc</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>b6785dc</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@fulghum __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>ad57645</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>ad57645</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

File: go/cmd/dolt/commands/cherry-pick.go [hunk 1]
```diff
@@ -154,7 +154,13 @@ hint: commit your changes (dolt commit -am \"<message>\") or reset them (dolt re
 		return fmt.Errorf("error: failed to set @@dolt_force_transaction_commit: %w", err)
 	}
 
-	q, err := dbr.InterpolateForDialect("call dolt_cherry_pick(?)", []interface{}{cherryStr}, dialect.MySQL)
+	interfaceArgs := make([]interface{}, 0, len(args))
+	for _, arg := range args {
+		interfaceArgs = append(interfaceArgs, arg)
+	}
+
+	placeholders := strings.Join(make([]string, len(args)), "?, ") + "?"
```
Comment (macneale4) [inline go/cmd/dolt/commands/cherry-pick.go:162]: Not wild about this. maybe a little helper function?

```
func generateQuestionMarks(count int) string {
    if count <= 0 {
        return ""
    }
    return strings.Repeat("?,", count-1) + "?"
}
```

`placeholders := generateQuestionMarks(len(args))`

We should probably have a helper function to convert arrays of strings into []interface{} instances. Kind of surprised I'm not finding that.

File: go/libraries/doltcore/cherry_pick/cherry_pick.go [hunk 2]
```diff
@@ -106,7 +142,11 @@ func CherryPick(ctx *sql.Context, commit string, options CherryPickOptions) (str
 		return "", nil, err
 	}
 	if pendingCommit == nil {
-		return "", nil, errors.New("nothing to commit")
+		if commitProps.SkipEmpty {
+			return "", nil, nil
```
Comment (macneale4) [inline go/libraries/doltcore/cherry_pick/cherry_pick.go:146]: the gDoc says that success will result in a hash (as a string). This seems to violate that. Probably worth calling out this edge case in the gDoc for the function.

File: go/libraries/doltcore/cherry_pick/cherry_pick.go [hunk 3]
```diff
@@ -299,6 +357,20 @@ func cherryPick(ctx *sql.Context, dSess *dsess.DoltSession, roots doltdb.Roots,
 	return result, cherryCommitMeta.Description, nil
 }
 
+func noChangesBetweenRoots(root1, root2 doltdb.RootValue) (bool, error) {
```
Comment (macneale4) [inline go/libraries/doltcore/cherry_pick/cherry_pick.go:360]: Is `rootsEqual()` a better name? Maybe a method on RootValue is needed? Seems odd that you need to go to all this effort to check that roots are equal

**Review by macneale4 — APPROVED** [review]
A couple minor questions that you can consider, but nothing to prevent shipping!

File: go/cmd/dolt/commands/cherry-pick.go (hunk 1 above)
Comment (fulghum) [reply go/cmd/dolt/commands/cherry-pick.go:162]: Good call, updated in 7f4810594a80d403e14dfc09cc3ac8c473f41b0c

File: go/libraries/doltcore/cherry_pick/cherry_pick.go (hunk 2 above)
Comment (fulghum) [reply go/libraries/doltcore/cherry_pick/cherry_pick.go:148]: Good catch. Updated in 3ee48c2eed346a5864307e6a4aa2d7e1340b36c7

File: go/libraries/doltcore/cherry_pick/cherry_pick.go (hunk 3 above)
Comment (fulghum) [reply go/libraries/doltcore/cherry_pick/cherry_pick.go]: Changed in 3ee48c2eed346a5864307e6a4aa2d7e1340b36c7

**Comment by coffeegoddd** [comment]
@fulghum __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>3ee48c2</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>3ee48c2</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Pull request merged**